mgr/dashboard: support inline edit for datatable

nizamial09 opened the pull request:
adding a new celltransformation that will transform the cell to a form
control when you click on the edit button which is inside a cell.

Added a new CellTemplate `editing` which if set to a cell, then it'll
add the edit button. You can also add validators to the control by using
the `customTemplateConfig` like
```
customTemplateConfig: {
          validators: [Validators.required]
        }
```

You can add the validations as part of the formGroup and its subsequent control.


Also using a `EditState` to keep track of the different cells I can edit simultaneously in a single time.


https://github.com/user-attachments/assets/554c9ede-81fb-49e9-aa73-2a9627de8be0

Fixes: https://tracker.ceph.com/issues/72171





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/icon/icon.component.ts:
```diff
@@ -7,12 +7,12 @@ import { ICON_TYPE, Icons, IconSize } from '../../enum/icons.enum';
 })
 export class IconComponent implements OnInit {
-  @Input() type!: keyof typeof ICON_TYPE;
+  @Input() type!: keyof typeof ICON_TYPE | keyof typeof Icons;
```
yeah, i thought of adding it to `ICON_TYPE` which meant I had to create another variable to instantiate the icon constants in the datatable and then use it for the type i want like

```
iconType = ICON_TYPE

<cd-icon [type]="iconType.check">
```

But since it already was using `icons = Icons` for other places inside the datatable, I thought it might be less work to keep on using enums as well and we can use the constants for more defined types like `success`, `errors` etc? WDYT?

afreen23 replied:
No you dont need to just add that here like this. And you wont run into any issues. :)
Add here `src/app/shared/enum/icons.enum.ts`

```
export const ICON_TYPE = {
  check: 'checkmark',
  copy: 'copy',
  danger: 'danger',
  infoCircle: 'info-circle',
  success: 'success',
  warning: 'warning'
} as const;

```
Then if any css for that icon (globally required) add here like `check-icon` in `src/app/shared/components/icon/icon.component.scss`

nizamial09 replied:
updated

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html:
```diff
@@ -396,3 +396,60 @@
+        #formDir="ngForm">
+    <div cdsRow>
+      <div cdsCol>
+        <cds-text-label [invalid]="column?.customTemplateConfig?.formGroup.controls[column?.prop].invalid && column?.customTemplateConfig?.formGroup.controls[column?.prop].dirty"
```
```suggestion
        <cds-text-label [invalid]="showError(column?.customTemplateConfig?.formGroup.controls[column?.prop])"
```

nizamial09 replied:
actually you shouldn't use function calls in angular template. 
https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html:
```diff
@@ -396,3 +396,60 @@
+                 size="sm"
+                 [id]="column?.prop"
+                 [formControlName]="column?.prop"
+                 [invalid]="column?.customTemplateConfig?.formGroup.controls[column?.prop].invalid && column?.customTemplateConfig?.formGroup.controls[column?.prop].dirty">
```
same as above

afreen23 commented:
@nizamial09 can you share an example for how to import this in a component ? As I guess this is needed in list component (the table cells) and we are adding form state ? Thanks

nizamial09 commented:
> @nizamial09 can you share an example for how to import this in a component ? As I guess this is needed in list component (the table cells) and we are adding form state ? Thanks

@afreen23 you have to create formgroup in the list file with the form controls you want and then add it to the table column like
```
name: 'name',
prop: 'name',
cellTransformation: CellTemplate.editing,
      customTemplateConfig: {
         formGroup: this.formGroup
      }
  ```
  
  The only thing is that the name of the control should match the prop. But I realize now that, that's not a good idea (because the prop can be something like `resp.id` and control might break?) but I'll have a look at that.

nizamial09 commented:
> But I realize now that, that's not a good idea (because the prop can be something like `resp.id` and control might break?) but I'll have a look at that.

i am leaving this as it is to make things minimal. Most of the time we use the prop without a `dot` so we might be okay. But if any special need arises we can take care of that later (we need to introduce an extra param for the form control name...)

nizamial09 commented:
jenkins test make check arm64

nizamial09 commented:
jenkins test windows

nizamial09 commented:
jenkins test dashboard

Pegonzal reviewed (commented):
Great addition, thanks @nizamial09 !

I was trying to test it in `rbd-list.component.ts`, I'm doing:
```
  formGroup = new CdFormGroup({
+    name: new FormControl('')
+  });
+

        name: $localize`Name`,
         prop: 'name',
         flexGrow: 2,
+        cellTransformation: CellTemplate.editing,
+        customTemplateConfig: {
+          formGroup: this.formGroup
+        },
```
I can see the inline edit, but it's not submitted, value is not udpating. What am I doing wrong?

[screen-capture (23).webm](https://github.com/user-attachments/assets/3aeb425d-d9c4-45b4-b237-dd98ab0b8afb)



Also, one thing I don't like about this, is that if you have many editable items there are a lot of edit icons, making the table too crowded IMO. 
Looking at the [carbondesign example in the docs](https://carbondesignsystem.com/patterns/common-actions/#edit) the request from my side, would be to have the icons hidden by default and make them visible on the item hover only, what do you think?

Another feedback from my side: could be allow to submit the change by pressing enter or by clicking outside the field? Personally I find that behaviour natural

nizamial09 commented:
> value is not udpating. What am I doing wrong?

that's because we are not consuming the button event that is emitted. The design is to emit an event when the button is clicked. And that event should be handled in the component (in this case, RBD List or in any of the place the event is used). Once that logic is set and we properly call the back-end this will be updated. The code i pasted is just to demo the front-end working.

> Looking at the [carbondesign example in the docs](https://carbondesignsystem.com/patterns/common-actions/#edit) the request from my side, would be to have the icons hidden by default and make them visible on the item hover only, what do you think?

maybe Ashwin should be the one deciding this. I'll check with him

> by pressing enter or by clicking outside the field?

pressing enter sounds legit but clicking outside might not be fine IMHO. I would still like them to click submit explicitly for changes to take effect. But again, I'll ask Ashwin to have a look.

ashwinjadhavIBM commented:
> > value is not udpating. What am I doing wrong?
> 
> that's because we are not consuming the button event that is emitted. The design is to emit an event when the button is clicked. And that event should be handled in the component (in this case, RBD List or in any of the place the event is used). Once that logic is set and we properly call the back-end this will be updated. The code i pasted is just to demo the front-end working.
> 
> > Looking at the [carbondesign example in the docs](https://carbondesignsystem.com/patterns/common-actions/#edit) the request from my side, would be to have the icons hidden by default and make them visible on the item hover only, what do you think?
> 
> maybe Ashwin should be the one deciding this. I'll check with him
> 
> > by pressing enter or by clicking outside the field?
> 
> pressing enter sounds legit but clicking outside might not be fine IMHO. I would still like them to click submit explicitly for changes to take effect. But again, I'll ask Ashwin to have a look.

Values should be saved only when the user presses Enter or clicks the Tick button. If they just type something and click outside the field, nothing should change and it should stay as it is.

nizamial09 commented:
@Pegonzal updated the PR

[Ceph Dashboard_ Block _ Images (1).webm](https://github.com/user-attachments/assets/a32247a7-726d-40f5-9c89-3ecb76348edf)

nizamial09 commented:
jenkins test api

nizamial09 commented:
jenkins test api

Pegonzal commented:
@nizamial09 I'm trying to test it, doing the same as last time in the rbd list with:

```
  formGroup = new CdFormGroup({
    name: new FormControl('')
  });
```

```
      {
        name: $localize`Name`,
        prop: 'name',
        flexGrow: 2,
        cellTemplate: CellTemplate.editing,
        customTemplateConfig: {
          formGroup: this.formGroup
        }
      }
```

But the Dashboard is breaking and I see the following error in the console: 

```
Uncaught TypeError: templateRef.createEmbeddedViewImpl is not a function
    at ViewContainerRef.createEmbeddedView (core.mjs:17109:37)
    at NgTemplateOutlet.ngOnChanges (common.mjs:4068:46)
    at NgTemplateOutlet.rememberChangeHistoryAndInvokeOnChangesHook (core.mjs:4114:14)
    at callHookInternal (core.mjs:5154:14)
    at callHook (core.mjs:5185:9)
    at callHooks (core.mjs:5135:17)
    at executeInitAndCheckHooks (core.mjs:5085:9)
    at refreshView (core.mjs:13792:21)
    at detectChangesInView (core.mjs:13993:9)
    at detectChangesInViewIfAttached (core.mjs:13953:5)
```

Any idea?

nizamial09 commented:
> This is just what I see when applying the changes I mentioned above:

tried again. didn't see the issue. I suspect it could be a dev server issue? can you try on a new cluster or maybe start a new cluster with the change.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:
```diff
@@ -384,6 +391,9 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
   private previousRows = new Map<string | number, TableItem[]>();
 
+  editingCells = new Set<string>();
+  editStates: EditState = {};
```
oh also one more important thing that I forgot, (which thankfully I added in the commit description) is 

> Also using a EditState to keep track of the different cells I can edit simultaneously in a single time.

I completely forgot about it. So yeah its necessary

nizamial09 commented:
@Pegonzal @afreen23 I've updated the PR again after seeing some issues now. And as part of it I've removed the part where you need to add a form control in the child table component. So you just have to set the `cellTransformation: CellTemplate.editing` and this will create formControls dynamically in the table component.

So now its just
```
      {
        name: $localize`Name`,
        prop: 'name',
        flexGrow: 2,
        cellTemplate: this.removingStatTpl,
        cellTransformation: CellTemplate.editing
      },
```
and optionally validators like
```
      {
        name: $localize`Name`,
        prop: 'name',
        flexGrow: 2,
        cellTemplate: this.removingStatTpl,
        cellTransformation: CellTemplate.editing,
        customTemplateConfig: {
          validators: Validators.required
        }
      },
```

So you can try again and let me know.
